### PR TITLE
Fixed columns creation and legend issue, now allows conditional coloring out-of-the-box

### LINKED ISSUE #202: y Axis title / label are not positionned properly on first load
When I load my graph for the first time, here is what is displayed : 
screenshot from 2015-03-14 17 08 14

And the if I click one of the titles to hide the serie and click again, the titles get spaces properly like that : 
screenshot from 2015-03-14 17 08 27

Any idea where this could come from.
Here is the conf of my graph : 

    {
      drawLegend: true,
      drawDots: true,
      stacks: [],
      axes: {
        x: { key: 'x', labelFunction: function(value) { return moment(value*1000).format('DD-MM-YYYY'); },
          type: 'linear', min: dataTot.xMin, max: dataTot.xMax, ticks: nbTicks},
        y: { type: 'linear', min: dataTot.yMin, max: dataTot.yMax, ticks: nbTicks}
      },
      series: [
        { y: 'y',  color: '#88BF57', thickness: '2px', type: 'line', dotSize: 3, label: 'Total' },
        { y: 'y2', color: '#414E92', thickness: '2px', type: 'line', dotSize: 3, label: 'Men' },
        { y: 'y3', color: '#EC7161', thickness: '2px', type: 'line', dotSize: 3, label: 'Women' }
      ],
      lineMode: 'monotone',
      tension: 0.7,
      tooltip: { mode: 'scrubber', formatter: function(x, y) { return moment(x * 1000).format('DD-MM-YYYY') + ' : ' + y; }},
      columnsHGap: 5
    }
    
Thanks a lot for the work by the way, very handy to create nice charts !

### LINKED ISSUE #217: Column color
Hi! Could you please help me to change color for different values of same column. For example
i need to use red, green, yellow if column has values 1,2,3 ?

### PR DESCRIPTION
**Pull request by chaosmail**
Hi n3-charts team,

I cleaned the columns code a little, in order to use the fill-attr (color option) directly on the rect node. This immediately allows conditional coloring by simply using a coloring function instead of a string as color value and fixes #217.

```js
var colorFn = function(d, i){
  return d && d.y > 10 ? 'red' : 'green';
};

$scope.options = {
  series: [
     {y: 'value', color: colorFn, type: 'column'}
  ]
};
```

Here is a demo: [http://plnkr.co/edit/0vDTzF6mUfDr1Tbnjzvk](http://plnkr.co/edit/0vDTzF6mUfDr1Tbnjzvk?p=preview)

This PR also fixes #202 some annoying bug where the legend labels are overlapping.

Best,
Christoph

### CODE REVIEW

File: .ideas.md [hunk 1]
```diff
@@ -1,3 +1,16 @@
+# Roadmap to 2.x
+
++ Data option should allow multiple separate serie
++ Axes labels
++ Grid
++ reusable class structure (would be especially handy for reusable stuff such as axes etc.)?
```
Comment (lorem--ipsum) [inline .ideas.md]: Please add a capital letter at the beginning of the sentence :-)

File: .ideas.md [hunk 2]
```diff
@@ -1,3 +1,16 @@
...
++ Don't always redraw the whole thing :) we can update just the changed parts
++ Separate tests into unit tests (with Karama) and Integration tests (with Protractor)
```
Comment (lorem--ipsum) [inline .ideas.md]: *Karma*

File: .ideas.md (hunk 1 above)
Comment (chaosmail) [inline .ideas.md]: Haha you are picky ;D sure I will do it

File: src/utils/legend.coffee [hunk 3]
```diff
@@ -65,63 +76,63 @@
...
-            'class': (d, i) -> "legendText series_#{i}"
-            'font-family': 'Courier'
-            'font-size': 10
-            'transform': 'translate(13, 4)'
-            'text-rendering': 'geometric-precision'
-          )
-          .text (s) -> s.label || s.y
-
+        # we need to do call this once, to not make the legend text
+        # does not blink on every update
```
Comment (lorem--ipsum) [inline src/utils/legend.coffee]: Comment should read "We need to call this once, so the legend text does not blink on every update"

File: src/utils/legend.coffee [hunk 4]
```diff
@@ -65,63 +76,63 @@
...
-        items.attr(
-          'transform': (s, i) ->
-            if s.axis is 'y'
-              return "translate(#{left.shift()},#{dimensions.height-40})"
-            else
-              return "translate(#{right.shift()},#{dimensions.height-40})"
-        )
+        # now once again,
+        # to make sure, text width gets really! computed properly
+        setTimeout translateLegends, 0
```
Comment (lorem--ipsum) [inline src/utils/legend.coffee]: Please use parenthesis for method calls (this is one of the CoffeeScript's features I don't like ^^)

**Comment by lorem--ipsum** [comment]
Great code, however the tooltip should read the color function too. Instead it will default to no color for scrubber, and `steelblue` for axis mode. Could you fix that please ? Otherwise it looks good.

File: .ideas.md (hunk 1 above)
Comment (lorem--ipsum) [inline .ideas.md]: Eh, you know what they say about broken windows :-D

File: src/utils/legend.coffee (hunk 3 above)
Comment (chaosmail) [inline src/utils/legend.coffee]: Yeah, actually I realized just now, that you are redrawing the whole chart on nearly every update :dancers: so i need to set the legend label first (with the wrong position) and then move it to the correct position

Comment (chaosmail) [inline src/utils/legend.coffee]: otherwise the legend text blinks once, because it is removed and redrawn

File: src/utils/legend.coffee (hunk 4 above)
Comment (chaosmail) [inline src/utils/legend.coffee]: sure :)

**Comment by chaosmail** [comment]
I have seen that scrubber tooltips are black when we use a coloring function - I will fix this. But do you mean the tooltips should also change colors?

**Comment by lorem--ipsum** [comment]
Yeah, that'd be cool :-) Or we would have to specify a default color for the series if it has a color function, because otherwise the tooltip would either stay black or pick a random color...

**Comment by chaosmail** [comment]
the nice thing was, that with the restructuring of the columns code, conditional coloring could be achieved without effort - because you can plugin the function instead of the color value. If we want the same behavior for tooltips, I need to mimik this behavior - and call the coloring function manually, if there is one

**Comment by chaosmail** [comment]
i just took a look at the code; I need to call it anyway manually, so I can also apply the conditional colors :)

**Comment by chaosmail** [comment]
The demo link is still the same [http://plnkr.co/edit/0vDTzF6mUfDr1Tbnjzvk]( http://plnkr.co/edit/0vDTzF6mUfDr1Tbnjzvk) actually, it's a really nifty feature with the tooltips also changing the color :)

File: src/utils/scrubber.coffee [hunk 5]
```diff
@@ -86,6 +86,13 @@
 
           positions[index] = {index, x: x, y: axes[v.axis + 'Scale'](v.y + v.y0), side, sizes}
 
+          # Use a coloring function if defined, else use a color string value
+          color = series.color?(v, series.values.indexOf(v)) ? series.color
```
Comment (lorem--ipsum) [inline src/utils/scrubber.coffee]: Could you use a more explicit syntax here ? Like `color = if angular.isFunction(series.color) then series.color(v, series.values.indexOf(v)) else series.color` or something like that ?

File: src/utils/tooltips.coffee [hunk 6]
```diff
@@ -106,8 +106,11 @@
         label = xTooltip.select('text')
         label.text(textX)
 
+        # Use a coloring function if defined, else use a color string value
+        color = series.color?(datum, series.values.indexOf(datum)) ? series.color
```
Comment (lorem--ipsum) [inline src/utils/tooltips.coffee]: Same here

File: src/utils/tooltips.coffee [hunk 7]
```diff
@@ -144,8 +147,11 @@
           'width': w
         )
 
+        # Use a coloring function if defined, else use a color string value
+        color = series.color?(datum, series.values.indexOf(datum)) ? series.color
```
Comment (lorem--ipsum) [inline src/utils/tooltips.coffee]: And here too

File: src/utils/tooltips.coffee [hunk 8]
```diff
@@ -164,9 +170,12 @@
           'w': w
         )
 
+        # Use a coloring function if defined, else use a color string value
+        color = series.color?(datum, series.values.indexOf(datum)) ? series.color
```
Comment (lorem--ipsum) [inline src/utils/tooltips.coffee]: And here too ^^

File: test/unit/column_series.mocha.coffee [hunk 9]
```diff
@@ -87,9 +87,12 @@ describe 'column series', ->
   it 'should create a group', ->
     content = element.childByClass('content')
     expect(content.children().length).to.equal 1
+    
     columnGroup = content.children()[0]
     expect(columnGroup.getAttribute('class')).to.equal 'columnGroup series_0'
-    expect(columnGroup.getAttribute('style').trim()).to.equal 'stroke: rgb(70, 130, 180); fill: rgb(70, 130, 180); fill-opacity: 0.8;'
+    
+    firstColumnStyle = columnGroup.children()[0].getAttribute('style').trim()
+    expect(firstColumnStyle).to.match /stroke: rgb\(70, 130, 180\); fill: rgb\(70, 130, 180\);(.)* fill-opacity: 0.7;/
```
Comment (lorem--ipsum) [inline test/unit/column_series.mocha.coffee]: Please use parenthesis

**Comment by lorem--ipsum** [comment]
Alright, almost there :-)

**Comment by chaosmail** [comment]
Hi Sébastien,
here is the rebased and corrected fix. I also added parenthesis to all equal and match calls in the unit tests.

Best,
Christoph

**Comment by lorem--ipsum** [comment]
:poultry_leg:

**Comment by rfcee** [comment]
Legend bug doesn't seem to be fixed, for graphs that are initially hidden:

http://plnkr.co/edit/PRSO0W9UUBGAjbHMpDpH?p=preview

**Comment by chaosmail** [comment]
@rfcee thanks for reporting this bug.

**Comment by rfcee** [comment]
When I comment out the line: 
setTimeout(translateLegends, 0);
...the legends do not overlap anymore. But when I do a window resize, then the legends overlap again. 

So I tried to comment out as well everything inside the window_resize function and just put:
scope.redraw;

Well that works, no legends overlapping, but the graphs don't resize :(

I hope someone can help me find a workaround for this...please.

**Pull request merged**